### ISSUE TITLE
Can an enumeration sort have one element?

### ISSUE THREAD

**Issue by endre90** [Z3Prover/z3#4953]
Hello,
I stumbled upon an error message today using the C API, and after a while I found a solution. However, I am still a bit interested why this happened. It turns out that when an enumeration sort has only one element, using the default smt solver in a non-incremental way breaks with:

```
ASSERTION VIOLATION
File: ../src/ast/datatype_decl_plugin.cpp
Line: 1224
UNREACHABLE CODE WAS REACHED.
Z3 4.8.10.0
Please file an issue with this message and more detail about how you encountered it at https://github.com/Z3Prover/z3/issues/new

```

Running the same instance with the default smt solver using push and pop breaks with: 
```
Segmentation fault (core dumped).
```
I also tried using the QF_FD solver, in a non-incremental way it again breaks with:
```
ASSERTION VIOLATION
File: ../src/ast/datatype_decl_plugin.cpp
Line: 1224
UNREACHABLE CODE WAS REACHED.
Z3 4.8.10.0
Please file an issue with this message and more detail about how you encountered it at https://github.com/Z3Prover/z3/issues/new

```
Running the same instance with the QF_FD solver using push and pop breaks with:
```
Error: Sorts (_ BitVec 2) and (_ BitVec 1) are incompatible
```

I tried to somehow recreate this error in Python, so I copied and modified the example from https://theory.stanford.edu/~nikolaj/programmingz3.html:

```
from z3 import *
s = SolverFor("QF_FD")
Color, (red, blue) = EnumSort('Color', ['red', 'blue'])
clr = Const('clr', Color)
s.add(clr == red)
print(s.check())
print(s.model())
```
which yields:
```
sat
[clr = red]
```
However, when I change the enumeration sort so that it has only one element:
```
from z3 import *
s = SolverFor("QF_FD")
Color, (red) = EnumSort('Color', ['red'])
clr = Const('clr', Color)
s.add(clr == red)
print(s.check())
print(s.model())
```
it breaks with:
```
yields: raise Z3Exception(msg)
z3.z3types.Z3Exception: Z3 expression expected
```
So in conclusion, in each case when the sort has more than one element, no errors occur. I'm filing this as an issue mostly because the error message returned by Z3 politely asked me to do so :)
However some insight would be nice :)
Cheers!

**Comment by NikolajBjorner** [comment]
There is some bug in QF_FD for enumeration sorts being translated into bit-vectors.
The issue with the python script is that it expects a list 

```
from z3 import *
s = SolverFor("QF_FD")
Color, [red] = EnumSort('Color', ['red'])
clr = Const('clr', Color)
s.add(clr == red)
print(s.check())
print(s.model())

```

**Comment by endre90** [comment]
Ok, thanks for the swift answer. I guess I will then close the thread for now.
